app/uk/gov/hmrc/transitmovements/controllers/MovementsController.scala

```diff
@@ -84,36 +91,93 @@ class MovementsController @Inject() (
 
   private def createArrival(eori: EORINumber): Action[Source[ByteString, _]] = Action.stream {
     implicit request =>
-      (for {
-        arrivalData <- movementsXmlParsingService.extractArrivalData(request.body).asPresentation
-        received = OffsetDateTime.ofInstant(clock.instant, ZoneOffset.UTC)
-        message <- messageFactory
-          .create(MessageType.ArrivalNotification, arrivalData.generationDate, received, None, request.body, MessageStatus.Processing)
-          .asPresentation
-        movement = movementFactory.createArrival(eori, MovementType.Arrival, arrivalData, message, received, received)
-        _ <- repo.insert(movement).asPresentation
-      } yield MovementResponse(movement._id, Some(movement.messages.head.id))).fold[Result](
-        baseError => Status(baseError.code.statusCode)(Json.toJson(baseError)),
-        response => Ok(Json.toJson(response))
-      )
+      withReusableSourceAndSize(request.body) {
+        (source, size) =>
```

Comment:
This is overdoing it a bit, because what's happening is that `Action.stream` is saving to a file, then this call is taking the file, reading it, then streaming it into another file. We should be able to fold this logic into `Action.stream` instead.